careers/devops.markdown

```diff
@@ -0,0 +1,33 @@
+---
+title: DevOps Engineer
+summary: The Haskell Foundation seeks a passionate DevOps professional to support GHC, related Haskell projects and tooling, and the HF itself.
+postedOn: 2021-12-28
+status: Open
+---
+
+# Job Details
+
+## Responsibilities
+
+* Maintain GitLab and CI infrastructure for GHC
+* Improve CI
+* Expand the CI infrastructure to include base the base Haskell ecosystem
+* Support other Haskell libraries and tools, including HLS (Haskell Language Server)
+* Work with the Haskell Infrastructure Team
+* Support OS packaging of GHC and tools
+* Documentation
+* Improvements and expansion of the GHC performance dashboard
+
+## Qualifications
+
+* Demonstrated ability to manage and simplify complexity
+* Collaborate with a variety of stakeholders
+* Can juggle multiple, conflicting priorities
+* Excellent communication skills
+
+The Haskell Foundation is looking for a DevOps engineer to take the Haskell ecosystem, primarily the Glasgow Haskell Compiler (GHC, the Haskell compiler), to the next level of stability, reliability, and performance.
+
+This role spans many Haskell ecosystem projects to homogenize and improve the infrastructure used to build GHC and other core open source projects whose maintainers are interested, for instance HLS, Cabal, Stack, GHCup, core libraries, and more.
+
+The successful candidate will be in the middle of it all, collaborating with developers at the forefront of making Haskell a groundbreaking, influential, and top tier programming language. The individual will improve existing infrastructure, work on creating a shared CI and performance benchmarking system across the Haskell ecosystem, shorted CI turnaround times, and improve debugability of failures.
+
```

Comment:
I'm left with some questions about the role. (I know the answers to some of these from other communication, but the answers should be in this posting.)

* Who would this person report to?
* Who would this person work with on a daily basis? What is the nature of the existing team?
* Where can we learn more about existing infrastructure?
* Does the applicant need to know Haskell?
* What specific technologies does the applicant need to know?
* What level of seniority are we looking for? Is there an experience requirement?
* Is there an education level requirement?
* Is the job on-site or remote?
* Are there geographical restrictions? Time zone preferences / restrictions?
* What's the HF's policy around flexibility in working conditions? What about time off? (While I don't think the job posting needs to have details here, having some idea of what to expect -- especially if the description is enticing -- may draw in more applicants.)
* What is the goal of the Haskell Foundation? What *is* the Haskell Foundation anyway? How is it funded?

Most job postings have some text around equity ("equal opportunity employer"). I'm not sure if this is legally required or just a good idea, but we should probably have such text.

Might also be good to link to the original proposal somewhere. That gives a nice dose of context and rephrases some of the job requirements (which may help readers understand better what's going on).

Reply:
hiring descriptions usually don't describe the answers to 1-3. i think there should be no hard exp or education requirements but obviously those qualifications will tend to matter, but there can be balances between one and the other as well.

Also the last one really can be determined by browsing... the entire rest of the site..

Reply:
I agree that job postings don't often contain much about 1-3, but I think clarifying these may be important here: there is currently 1 employee of the HF (and in a mostly non-technical role), and I would want an applicant to know they would be working with more than just that 1 employee (which an applicant might otherwise think). My hope is that working with several others might be seen as a good thing and encourage more applications. If it doesn't fit in the context of the rest of the text, I could see skipping this.

I'm agnostic as to any specific education or experience requirements -- I see advantages and disadvantages in every direction. But it might be good to say what the requirements are (or that we don't have any), lest an applicant erroneously think "I don't have a master's in CS, so I can't apply".

As for the last question: that was inspired by looking at several job descriptions out there for similar positions, which do tend to have a few sentences introducing the company (even when that company is very well known). I don't feel strongly (because an applicant can easily find the information), but it did seem to be part of several other postings.

Reply:
> Who would this person report to?

It is pretty important to mention properly, as this is a specific context and an attractor. That the person can mostly be independent is a high attraction for professionals, DevOps'es maybe love to be even more antisocial than Devs. The "ability to work independently/autonomous" is at least good to mention. But, yes - an aware DevOps needs a ruling organ to be able to go and:
  1. report difficult problems that need money or specific specialist, or negotiate migration
  2. determine priority during  "juggle and choose among multiple, conflicting priorities" - the person needs to be mostly independent, but in overloaded situations DevOps needs to notify and establish hard choices prioritization with management,
  3. escalate the problem
  4. - important - gather the vies of the end customers & probe & use their value system.

In the original idea of "agile" is for engineering to be oriented on the value system of the customer. For DevOps - the end customer is Haskell engineers, so DevOps needs to have the freedom to discuss minor things in the chatroom and for the Haskell committee of engineers to be open to get the input questions from DevOps & give the output, as smart architect/DevOps would rely on the "terminal device" value-design system and understanding as a basis for design. Haskell DevOps needs to be a member of the group, to attend the meetings to I/O information on design and understanding.

> What specific technologies does the applicant need to know?

Is one of the most important for DevOps people. DevOps job posting without listing of technologies just reads/looks weird, as there is always happy/useful information to transmit through the enumeration of tech.

DevOps'es strongly review/judge/infer work by a stack of technologies mentioned. As DevOps has a lot of tools / DSLs - the listing has enough information to infer a lot about the work.
  * Current infrastructure listing of terms (so DevOps knows the starting point)
  * Imaginary desirable infrastructure listing to arrive to (to select people who know those technologies & can/motivated to arrive there & show that you give them technological futurism/freedom to construct infrastructure)
    * Terraform and containerization are probably worth mentioning in the listing as a desirable end point in any case.
    * Despite complexity & difficulties in the Cabal packaging - it is structurally isolated from the DevOps infrastructure. That architectural boundary needs to be revealed.
      * Probably, portion by portion Haskell DevOps infrastructure can be gradually migrated to the Nix stack.
      * From experience can tell that the person that knows Haskell - understands/knows/easily learns Nix better - and would learn, know & write a better Nix code than a person which learned Nix only. So the person that both knows Haskell & Nix somewhat - (with the caveat of packaging design boundary, otherwise) has mutual synergy with Haskellers & becomes a better Nix'er.
      * Usage of NixOS in the long term would save DevOps maintenance efforts & infrastructure.
      * Nix establishes a strong guarantee/requirement that almost all infrastructure would be saved as an infrastructure as a code. Person does not need to be controlled - Nix requires DevOps to be done as infrastructure as a Nix code.
      * As a result, in any situations infrastructure as a code would be easy to hand over.
      * Would be easy to add people into the Ops team, easy to extend it with "scratch own itch" volunteer flow.
      * It would be easy to inspect projects and infrastructure & would be easy for the community to chime in with Pull Requests. So person would not be totalitarian-like alone with infrastructure, it would be a community process.
      * NixOS/Nixpkgs community & expertise is vast.
      * A lot of people want to work in Nix DevOps
      * NixOS/Nixpkgs DevOps talent pool is not tapped into by business. It is a rich and motivated talent pool.
      * The NixOS/Nixpkgs DevOps talent pool has a community collaboration value system that aligns with what Haskell Foundation nature.
      * So in NixOS/Nixpkgs community - a great number of experienced DevOps'es can be found & eager to work in Haskell Foundation.
      * The positive community development feedback loop is strong.
      
Overall the current stack needs to be described, and the second listing is the fictional end stack as an example.

Reply:
If Nix would be considered - we togather happen to know/be connected to/have a network of a lot of Nix DevOps mastodonts. It would be immensely useful to send them information directly.

Reply:
This person will (to my knowledge) report to @bgamari.

This is a very versatile role.  The tech stack is a mixture of nix (some PS1 scripts on windows) and gitlab. One of the key points to consider is that we run on a lot of donated hardware, so access is not always trivial (this depends a lot of building up trust) and might end up needing to be communicated with third party providers. If someone brings in some Haiku builder, I'm fairly sure we can't use nix on that.

Core skills to me are
- excellent analytical
- excellent social
- good ops experience
- good haskell knowledge (you don't need expert haskell knowledge, but you need to at least be able to understand why some thing might fail. We have haskell engineers to fix bugs, the CI is *not* written in haskell).
- gitlab experience preferably 
- data collection, visualisation and interpretation experience would be good
- optional macOS experience (we have lots of people comfortable with linux, but debugging macOS issues not so many)

Now the tech stack is not completely fixed, but we do build on linux, macOS, windows, bsd, and various architectures. Some can be managed by nix, some can't.

This is the core CI part. Issues that come up are mostly: some builder starts acting up, what's wrong, why does that builder behave strange?

Collect performance metrics of builds, and figure out if we can improve performance.

Reply:
> This person will (to my knowledge) report to @bgamari.

This surprises me. For me, the "reports to" relationship implies that e.g. @bgamari would have the authority to fire the devops person, likely in consultation with others. It also implies that when the hire is ready to ask for a raise, they would ask @bgamari -- or that @bgamari would have a definitive say on the matter. It would be unusual (to me) to have that relationship fall outside the HF. I'm not saying this is problematic, per se, but it does strike me as unusual and something that should be explicitly agreed to on all sides.

Reply:
https://github.com/haskellfoundation/haskellfoundation.github.io/pull/169#discussion_r782678617 sounds to me, as I pointed out "in any case", because indeed Nix can not be/sometimes is not effective to be used - so the use of Terraform, Docker & Ansible is required.

Reply:
If a person reports to CTO, it does not mean that CTO is who decides to fire/hire a person or decides on monetary spending. The money/fire/hire questions are generally fit for collective/hybrid decision structure, while it is also nearly mostly impossible to report/have as a direct order from the collective, because for that to function - it implies a requirement of quorum from a collective to agree on decisions. While DevOps frequently has a high level of requirement for stochastic reporting & negotiation. Technical bug, hardware, DDoS, or service - DevOps may be required to do sudden work with hard choices & the chief negotiation then is needed, & nobody would collect a group of people fast enough for that, so having 1 person to report to is needed.

Reporting/decision making from a collective requires an overhead from the collective & the other part - is that in the collective the responsibility gets split while on requests/dialog through time certain parts of a collective can be unaware of prior decisions & give contrary decisions/responses.

The etymology of "manager" alludes to why that point in structure exists & why we keep semantically meaning of managing people.

It is mostly desirable for the 1 formal I/O entry point to exist/be determined, and for people to build relationships and understanding between each other & establish the style of the work process.

`design/todo` flow is a different process from hiring/money decision-making. The hiring/money questions are generally ones that escalate way beyond the worker-manager, as, I believe, HF already has a monetary organ for example.

Reply:
Indeed I would hope that there would be a contact within HF who would be responsible for "HR" relevant interactions. However, as I've stated earlier I would be happy to help set tasks and priorities.

Reply:
This is a key bit:

> as I've stated earlier I would be happy to help set tasks and priorities

Thank you for doing this! However, there is (in my mind) a biggish difference between "help set tasks and priorities" and "set tasks and priorities". The former calls to mind regular checkins with Andrew, who assigns tasks, sometimes informed by Ben's technical details (say). The latter says that Ben chooses the tasks and assigns them. The current text reads like the latter, but Ben's response says the former.

I could be splitting hairs here. (Actually, I *am* splitting hairs. It's just that I think this is an important hair to split!) It just seems really easy to me to have miscommunication and mistaken expectations here, and a mismatch here would have a very long shadow.

Reply:
It is indeed true.

The wording of the application posting conveys a lot to the job seeker, as that is all information person often has about the work & climate in the company. Indeed, the simple word "help" implies a tremendous difference in reading & understanding, as it moves from the rigid two-way architectural dependency between worker & manager, into a lax one. From totalitarian into actor model.

The analog of difference in law or standards would be the `may` word.